docs/tutorials/mongoid-queries.txt

```diff
@@ -198,6 +198,93 @@ loaded with ``only`` for those associations to be loaded. For example:
   # => [#<Manager _id: 5c5dc2f0026d7c1730969843, band_ids: [BSON::ObjectId('5c5dc2f0026d7c1730969842')]>]
 
 
+Logical Operations
+******************
+
+Mongoid supports ``and``, ``or``, ``nor`` and ``not`` logical operations on
+``Criteria`` objects. These methods take one or more of
+a hash of conditions or another ``Criteria`` object as their arguments.
+As of Mongoid 7.1, logical operations have been changed to offer the
+the same semantics as `those of ActiveRecord
+<https://guides.rubyonrails.org/active_record_querying.html>`_.
+
+.. code-block:: ruby
+
+  Band.where(label: 'Trust in Trance').and(name: 'Astral Projection')
+  
+  Band.where(label: 'Trust in Trance').or(Band.where(name: 'Astral Projection'))
+  
+  Band.not({label: 'Trust in Trance'}, Band.where(name: 'Astral Projection'))
+
+and behavior
+````````````
+
+The ``and`` method will add new simple conditions to the top level of the
+criteria, unless the receiving criteria already has a condition on the
+respective fields, in which case the conditions will be combined with ``$and``.
+
+.. code-block:: ruby
+
+  Band.where(label: 'Trust in Trance').and(name: 'Astral Projection').selector
+  # => {"label"=>"Trust in Trance Records", "name"=>"Astral Projection"}
+
+  Band.where(name: /Best/).and(name: 'Astral Projection').selector
+  # => {"name"=>/Best/, "$and"=>[{"name"=>"Astral Projection"}]}
```

Comment:
Why doesn't this do:
```
 {"$and"=>[{"name"=>/Best/}, {"name"=>Astral Projection"}]}
```
See: https://docs.mongodb.com/manual/reference/operator/query/and/

Reply:
I tried this approach first but it runs into practical complications when chaining more than one time. The current behavior is:

<pre>
irb(main):004:0> Foo.where(bar1: 1).where(bar2: 2).where(bar3: 3)
=> #<Mongoid::Criteria
  selector: {"bar1"=>1, "bar2"=>2, "bar3"=>3}
  options:  {}
  class:    Foo
  embedded: false>
</pre>

Using `$and` between conditions would alter the resulting query for existing applications.

The approach in this PR produces the same behavior for different keys, and adds:

<pre>
irb(main):004:0> Foo.where(bar: 1).where(bar: 2).where(bar: 3)
=> #<Mongoid::Criteria
  selector: {"$and"=>[{"$and"=>[{"bar"=>1}], "bar"=>2}], "bar"=>3}
  options:  {}
  class:    Foo
  embedded: false>
</pre>

Reply:
I think we should **not** do this:

```
Foo.where(bar: 1).where(bar: 2).where(bar: 3)
=> selector: {"$and"=>[{"$and"=>[{"bar"=>1}], "bar"=>2}], "bar"=>3}
```

Reply:
Doing so requires inspecting the contents of conditions, whereas the current implementation can combine conditions without inspecting their contents.

Reply:
I think it is a noble goal, and certainly one that can be discussed once all of the operators are handled and tests are written. In the interests of making forward progress I focused on attaining the desired behavior with respect to condition combination first and leaving the visuals of constructed queries for later.

Reply:
I would not want to accept a PR here that doesn't preserve visual cleanliness of operators.

Reply:
I agree that, everything else being equal, a more visually appealing output is preferable, but can you please clarify whether you have a use case that depends on the shape of queries generated by Mongoid?

Reply:
Yes, absolutely, we have monitoring scripts which depend on the shape of certain queries in the Mongo logs.

In addition, ` selector: {"$and"=>[{"$and"=>[{"bar"=>1}], "bar"=>2}], "bar"=>3}` is very difficult for humans to read compared to `selector: {"bar1"=>1, "bar2"=>2, "bar3"=>3}`; it makes the whole library significantly harder to work with.

Lastly, have you confirmed that the two queries have identical performance, index behavior, etc in MongoDB itself?

Reply:
This is a fundamentally breaking change in my opinion which should not be allowed. It is CERTAINLY not worth the "nice-to-have" addition of AREL-style syntax.

Reply:
I think you are confusing multiple conditions on the same fields and conditions on different fields, because as I mentioned in an earlier comment this PR preserves existing behavior when conditions on different fields are combined.

Reply:
Ah, ok I see. But in this case, why not flatten the `$and` into:

```
irb(main):004:0> Foo.where(bar: 1).where(bar: 2).where(bar: 3)
=> #<Mongoid::Criteria
  selector: {"$and"=>[{"bar"=>1}, {"bar"=>2}, {"bar"=>3}]}
```

Reply:
This was my first thought as well, but again, once more complex queries are considered, plus the multitude of ways Mongoid permits the queries to be constructed, the choice of query shape becomes less obvious. The documentation attached to this PR will explain how the queries are produced.

Reply:
The query shape differences between where and and will be addressed in https://jira.mongodb.org/browse/MONGOID-4806.

Reply:
I think it should be done as part of this PR. Shouldn't introduce new strange behavior.